### ISSUE TITLE
Some in-memory statistics reporting twice their values

### ISSUE THREAD

**Issue by reyjrar** [graphite-project/carbon#716]
Relevant Version: 
   * graphite-web 1.0.2
   * carbon   1.0.2

I recently tuned the carbon-cache settings to lower disk thrashing.  This means I have metrics staying in the in-memory cache for longer than before, up to ~21 minute or so.  Most metrics return the correct values for their entire series.  However a group of metrics with the same metric prefix routinely return ~21 minutes of doubled data at the leading edge of the graph. 

You can see the oddity in the "Graphite Fetch Differences" graph below, there are two instances each mirrored and depending on how much of the data is flushed to disk, they return different values.  Immediately after a carbon-cache restart, the values are correct, but as more data builds up in the caches, the values are consistently doubled.

[See graph here since GitHub's Image is corrupted](http://edgeofsanity.net/media/Cache-odditiy.png)

I can't explain what's happening here, but my setup is: carbon-c-relay using the carbon hash ring to write to 8 carbon-cache local instances.  Graphite-web is configured to use the carbon hash ring in the same order, same keys to fetch results from the carbon-cache instances.

```
cluster caches
    carbon_ch
        127.0.0.1:2103=1
        127.0.0.1:2203=2
        127.0.0.1:2303=3
        127.0.0.1:2403=4
        127.0.0.1:2503=5
        127.0.0.1:2603=6
        127.0.0.1:2703=7
        127.0.0.1:2803=8
;
match * send to caches;
```
Carbon cache configs:
```
# <SNIP> 
# Cache Queues
# ------------------------------------------------------------------------
MAX_CACHE_SIZE                     = 83333333
MAX_UPDATES_PER_SECOND             = 200
MAX_UPDATES_PER_SECOND_ON_SHUTDOWN = 20000
MAX_CREATES_PER_MINUTE             = 500
USE_FLOW_CONTROL                   = True

# ------------------------------------------------------------------------
# Disk IO
# ------------------------------------------------------------------------
# CACHE_WRITE_STRATEGY:
#  sorted - default, metrics counted and flushed in order
#  max    - most frequently accessed items flushed first
#  naive  - best for solid state storage, no sort pressure on carbon-cache
CACHE_WRITE_STRATEGY     = naive
WHISPER_FALLOCATE_CREATE = True
WHISPER_LOCK_WRITES      = True

# ------------------------------------------------------------------------
# Listen Interfaces
# ------------------------------------------------------------------------
ENABLE_UDP_LISTENER       = False
LINE_RECEIVER_INTERFACE   = 127.0.0.1
PICKLE_RECEIVER_INTERFACE = 127.0.0.1
CACHE_QUERY_INTERFACE     = 127.0.0.1

# ------------------------------------------------------------------------
# Cache Instances
# ------------------------------------------------------------------------
[cache:1]
LINE_RECEIVER_PORT=2103
PICKLE_RECEIVER_PORT=2104
CACHE_QUERY_PORT=7102
[cache:2]
LINE_RECEIVER_PORT=2203
PICKLE_RECEIVER_PORT=2204
CACHE_QUERY_PORT=7202
[cache:3]
LINE_RECEIVER_PORT=2303
PICKLE_RECEIVER_PORT=2304
CACHE_QUERY_PORT=7302
[cache:4]
LINE_RECEIVER_PORT=2403
PICKLE_RECEIVER_PORT=2404
CACHE_QUERY_PORT=7402
[cache:5]
LINE_RECEIVER_PORT=2503
PICKLE_RECEIVER_PORT=2504
CACHE_QUERY_PORT=7502
[cache:6]
LINE_RECEIVER_PORT=2603
PICKLE_RECEIVER_PORT=2604
CACHE_QUERY_PORT=7602
[cache:7]
LINE_RECEIVER_PORT=2703
PICKLE_RECEIVER_PORT=2704
CACHE_QUERY_PORT=7702
[cache:8]
LINE_RECEIVER_PORT=2803
PICKLE_RECEIVER_PORT=2804
CACHE_QUERY_PORT=7802
```

Graphite-web config:
```
# ------------------------------------------------------------------------
# Configure Carbon Query Link
# ------------------------------------------------------------------------
CARBONLINK_HASHING_TYPE = 'carbon_ch'
CARBONLINK_HOSTS = ["127.0.0.1:7102:1", "127.0.0.1:7202:2", "127.0.0.1:7302:3", "127.0.0.1:7402:4", "127.0.0.1:7502:5", "127.0.0.1:7602:6", "127.0.0.1:7702:7", "127.0.0.1:7802:8"]
CARBONLINK_TIMEOUT    = 5.0
CARBONLINK_QUERY_BULK = True
```

**Comment by deniszh** [comment]
That's an odd issue, I didn't experience anything similar. Anyone else?

**Comment by deniszh** [comment]
What's the functions are applied on the affected graph, @reyjrar ?

**Comment by reyjrar** [comment]
It's a sumSeries() across a key space of about 25 metrics.

Here's some more information:  the "shared" prefix is my statsite prefix.

In order to allow rebooting of statsd servers for upgrades and maintenance, my clients statsd to 2 servers.  So every one of these metrics is being sent twice.  Given that the servers aren't restarted at precisely the same time, the updates may come in splayed a bit so:

client ---- statsite-1 ------> | graphite |
           \--- statsite-2------> |   cluster  |

**Comment by deniszh** [comment]
How often do you sending these metrics?
Maybe cache retention is no the same as whisper file retention (just thinking out loud)

**Comment by reyjrar** [comment]
They're being sent every 60s from the statsite daemon to the carbon-caches, this matches the storage retention on the disk.

**Comment by deniszh** [comment]
I'll try to reproduce that.

**Comment by reyjrar** [comment]
Well, I shutdown the redundant statsite forwarder for 20 minutes, no dice. :(  Still saw the leading edge values doubled.

**Comment by cbowman0** [comment]
I wonder what's in the carbon caches.  Can you query them to find out what the contents look like for the metrics?

I use something like this when I'm debugging data: https://gist.github.com/cbowman0/c0eeadd59e5b239c1ed1a1e4605bf736
Run it from the server and it'll pick up the graphite configs.  Pass the key to query.  Example:
```
# /opt/graphite/bin/query_carbon.py metrics.a.b.c.d.key
[   (1513990020.0, 62.626483644730925),
    (1513990080.0, 62.719205811640585),
    (1513990140.0, 62.634264458874895),
    (1513990200.0, 62.32066779789308),
    (1513990260.0, 62.56138712390221),
    (1513990320.0, 62.413503417269126),
    (1513990380.0, 62.40627817107656),
    (1513990440.0, 62.08438769836514),
    (1513990500.0, 62.32332876368132),
    (1513990560.0, 62.41397099459579),
    (1513990620.0, 62.424858660789525),
    (1513990680.0, 62.22668075700878),
    (1513990740.0, 62.31362078107886),
    (1513990800.0, 62.09253780327061),
    (1513990860.0, 62.768599550709446),
    (1513990920.0, 62.764459349123534),
    (1513990980.0, 62.99036626026032),
    (1513991040.0, 62.81014887199259),
    (1513991100.0, 63.03978468983039),
    (1513991160.0, 63.02386159258445),
    (1513991220.0, 63.26557829215759),
    (1513991280.0, 63.057384279002655),
    (1513991340.0, 62.91130099861936),
    (1513991400.0, 62.678036543970435)]
```

**Comment by deniszh** [comment]
Nice tool, @cbowman0 !
Maybe we should include it somewhere in contrib.

**Comment by reyjrar** [comment]
Unsuccessful, I just get:

```
Traceback (most recent call last):
  File "./carbon_fetch.py", line 20, in <module>
    pp.pprint(CarbonLink().query(metric))
AttributeError: CarbonLinkPool instance has no __call__ method
```

For:
```
#!/usr/local/bin/python2.7

import sys
sys.path.append('/use/share/graphite/webapp')

import os
os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'graphite.settings')

import graphite.settings
from graphite.carbonlink import CarbonLink

import pprint
pp = pprint.PrettyPrinter(indent=4)

if (len(sys.argv) < 2):
  print "Pass metrics to query"
  sys.exit(1)

metric = sys.argv[1]
pp.pprint(CarbonLink().query(metric))
```

Had to tweak the script since I'm not installing into /opt.

**Comment by cbowman0** [comment]
Doh.  Carbonlink interface changed from 1.0.2.  Need to drop the `()`.  Here's an updated version: https://gist.github.com/cbowman0/32b0fe59be7d5d4228be902e2e7aef19

**Comment by deniszh** [comment]
Did you check what carbonlink returns BTW, @reyjrar ?
Just curious did you find an issue.

**Comment by reyjrar** [comment]
I just cycled back to this.  When I run the tool @cbowman0 pasted for any of the metrics returning double the actual value, the numbers are correct.  However, looking at the same data in graphite-web, I'm seeing the in-memory statistics doubled.  This might be related to https://github.com/graphite-project/graphite-web/issues/1988 ??

**Comment by deniszh** [comment]
>When I run the tool @cbowman0 pasted for any of the metrics returning double the actual value, the numbers are correct.

Are you writing statsd metrics on different nodes simultaneously? 
Could you please check data from same metrics but on different nodes of cluster?

>This might be related to graphite-project/graphite-web#1988 ??

Unfortunately, https://github.com/graphite-project/graphite-web/issues/1988 is a big mistery for me too.
I'm suspecting that bug lies somewhere in merging code, and probably related with a fact that carbonlink data is bit different on 2 nodes of the cluster.

**Comment by reyjrar** [comment]
I just completed the upgrade to 1.1.1 today.  So far there's no metric doubling via graphite-web, so it might be fixed in 1.1.1

**Comment by reyjrar** [comment]
This hasn't recurred, closing the issue.
